### PR TITLE
ECDSA alternative support

### PR DESCRIPTION
**Pull request by RonEld**
## Description
Support for alternative implementation of ECDSA, at the higher layer.
Doesn't include sw fallback support

## Status
**READY**

## Requires Backporting
NO  

## Migrations
If there is any API change, what's the incentive and logic for it.

YES
adding support for alternative implementation of ECDSA.

## Additional comments
The only supported alternative API is `mbedtls_ecdsa_sign`, `mbedtls_ecdsa_verify` and `mbedtls_ecdsa_genkey`.
I didn't see a reason for other API to be alternative implemented, because the rest of the APIs receive `mbedtls_ecdsa_context` which is treated as the key pair outside the APIs as well. The APIs that were selected for alternative implementation are the lowest APIs that receive the keys as `mbedtls_mpi` and `mbedtls_ecp_point` , which are specifically the keys for the operations. Although `mbedtls_ecdsa_genkey` receives `mbedtls_ecdsa_context` as parameter, it is an output parameter, and HW PKI are often used for generating the key pairs.

## Todos
- [x] Tests
- [ ] Documentation
- [x] Changelog updated
- [ ] Backported

## Steps to test or reproduce
Can be tested only with alternative implementation of ECDSA

### CODE REVIEW

File: library/ecdsa.c [hunk 1]
```diff
@@ -37,11 +37,10 @@
 #include "mbedtls/asn1write.h"
 
 #include <string.h>
-
+#include "mbedtls/platform.h"
```
Comment (andresag01) [inline library/ecdsa.c]: Question: Why did you add this `#include` here?

Comment (RonEld) [reply library/ecdsa.c]: I think it's by mistake, when I added prints for debugginig. I'll remove

**Comment by RonEld** [comment]
Hi @andresag01 I removed the redundant include

File: library/ecdsa.c [hunk 2]
```diff
@@ -37,11 +37,9 @@
 #include "mbedtls/asn1write.h"
 
 #include <string.h>
-
 #if defined(MBEDTLS_ECDSA_DETERMINISTIC)
 #include "mbedtls/hmac_drbg.h"
 #endif
-
```
Comment (andresag01) [inline library/ecdsa.c]: I think this change and the one in line 40 are only style so not directly related to this PR...

File: ChangeLog [hunk 3]
```diff
@@ -36,6 +36,10 @@ Changes
    * Clarify ECDSA documentation and improve the sample code to avoid
      misunderstandings and potentially dangerous use of the API. Pointed out
      by Jean-Philippe Aumasson.
+   * Add support for alternative implementation for ECDSA, controlled by new
+     configuration flag MBEDTLS_ECDSA_ALT in config.h.
+     Alternative Ecdsa is supported for implementation of `mbedtls_ecdsa_sign`
```
Comment (andresag01) [inline ChangeLog]: Style: Ecdsa -> ECDSA

File: ChangeLog [hunk 4]
```diff
@@ -36,6 +36,10 @@ Changes
...
+     and `mbedtls_ecdsa_verify`.
```
Comment (andresag01) [inline ChangeLog]: I think `mbedtls_ecdsa_genkey()` is also supported

Comment (andresag01) [reply ChangeLog]: Suggested rephrasing:
>Alternative Ecdsa is supported for implementation of `mbedtls_ecdsa_sign` and `mbedtls_ecdsa_verify`.

to
>The following function from the ECDSA module can be replaced with an alternative implementation: mbedtls_ecdsa_sign(), mbedtls_ecdsa_verify() and mbedtls_ecdsa_genkey().

File: library/ecdsa.c [hunk 5]
```diff
@@ -301,7 +314,6 @@ static int ecdsa_signature_to_asn1( const mbedtls_mpi *r, const mbedtls_mpi *s,
 
     return( 0 );
 }
-
```
Comment (andresag01) [inline library/ecdsa.c]: Same as the other comment. THis is only a style change and not directly related to the PR

**Comment by RonEld** [comment]
Hi @andresag01 , I addressed your comments
Please approve if satisifed

**Comment by andresag01** [comment]
@RonEld: Many thanks for considering my review comments :)

File: include/mbedtls/config.h [hunk 6]
```diff
@@ -238,6 +238,7 @@
 //#define MBEDTLS_BLOWFISH_ALT
 //#define MBEDTLS_CAMELLIA_ALT
 //#define MBEDTLS_DES_ALT
+//#define MBEDTLS_ECDSA_ALT
```
Comment (yanesca) [inline include/mbedtls/config.h]: You are providing option to replace three functions, therefore I think it would be more appropriate to use a MBEDTLS__FUNCTION_NAME__ALT type macro for each instead of a single MBEDTLS__MODULE_NAME__ALT type flag.

Comment (RonEld) [reply include/mbedtls/config.h]: I understand your point, and I will work on it, although I don't see any other way to have an alternative implementation for ECDSA, since all other functions are related to ASN1 decoding, and other modules not related specifically to the cryptographic operation

Comment (yanesca) [reply include/mbedtls/config.h]: I am not sure I understand. I was suggesting to do the same that you already do, but instead of using a single `MBEDTLS_ECDSA_ALT` macro, use three different flags (`MBEDTLS_ECDSA_SIGN_ALT`, `MBEDTLS_ECDSA_VERIFY_ALT`, `MBEDTLS_ECDSA_GENKEY_ALT`) in the same way you already do, just slicing the guarded code segment to 3 disjunct pieces.

The result would be pretty much the same like for the `MBEDTLS_AES_XXX_ALT` macros, and these would go into the same section in `config.h`.

Would this raise any significant issues for the ECDSA alternative implementation?

Comment (RonEld) [reply include/mbedtls/config.h]: It wouldn't raise significant issues. What I mean is, that as I see it, and alternative implementation, should implement only these three functions( perhaps without the genkey, and this is why I will do as you suggested), but personally I don't like adding to much configurable options, as it increases maintenence

Comment (yanesca) [reply include/mbedtls/config.h]: I completely agree that it would be more concise and simple to have only one flag, and I see the benefit of this too. My problem is, that it wouldn't be consistent with our current model of alternative implementations (either replace a whole module or just a single function) and I would prefer to have consistency over simplicity here.

**Comment by RonEld** [comment]
Hi @yanesca I changed according to your comments
Please review

**Comment by sbutcher-arm** [comment]
Reviewed and approved for merge.

**Comment by Patater** [comment]
retest
